### ISSUE TITLE
`D102` permits missing docstrings in public methods if Python file name starts with `_`

### ISSUE THREAD

**Issue by yury-fedotov** [astral-sh/ruff#14219]
## Checks

List of keywords I searched for before creating this issue:
* D102

## Description

It seems like `D102` permits missing docstrings in public methods if Python **file name** starts with `_`. This probably should not be the case, since the rule name suggests that it is concerned only with **method names**.

## MRE

```python
# File is called "_foo.py" - underscore is important.

class MyClass:

    def __init__(self) -> None:
        pass
    
    def important_method() -> None:
        pass
```

Example above passes D102. If you rename the file to `foo.py` - it would fail.

## Run params

* The command you invoked: `ruff check`
* The current Ruff settings: D102 enabled.
* The current Ruff version: 0.7.1.

**Comment by charliermarsh** [comment]
Thanks for the nice write-up. I think this is the same as #14206?

**Comment by yury-fedotov** [comment]
> Thanks for the nice write-up. I think this is the same as #14206?

Yes, I think so. Interesting that two people reported that in a 10 hour interval lol 😄 . Feel free to close mine then?

**Comment by charliermarsh** [comment]
Haha yeah. It's funny when that happens -- we haven't touched this behavior in months!

**Comment by gaborbernat** [comment]
I don't think these two are exactly the same. In my request I would expect that the current behavior persists however it would change when something is enumerated within the all global. However here, I assume you just want all private modules, public entries to face this rule.
